### PR TITLE
Add torch.library.impl_device

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111310
* #111380
* __->__ #111309
* #111308
* #111307

Use this to register an implementation for a specific device_type.
We map the user provided device-type(s) to DispatchKeys and then call
torch.library.impl on them.

They may also specify multiple device_types at once or specify
"device_types=default" to get a CompositeExplicitAutograd registration.

Test Plan:
- new tests

### CODE REVIEW

File: torchgen/model.py [hunk 1]
```diff
@@ -85,6 +85,7 @@ class DispatchKey(Enum):
     SparseCsrCPU = auto()
     SparseCsrCUDA = auto()
     NestedTensor = auto()
+    Dense = auto()
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torchgen/model.py:88]: Before this PR, we already had the binding for torch._C.DispatchKey.Dense. We need to add a line here to add the type hint for it.

**Comment by zou3519** [comment]
> Can you tell me what the difference between impl and impl_device is? I know we did chat about this, but looking over this PR I am sure feeling mournful that you can't just use impl to mean impl_device

torch.library.impl is a pre-existing API that lets people register kernels for dispatch keys.
torch.library.impl_device is an API for registering kernels for specific device types.

People start doing weird things when you tell them about DispatchKeys, so torch.library.impl_device hides that from the user. {torch.library.defiine, torch.library.impl_abstract, and torch.library.impl_device} are a collection of higher-level APIs (that we'll likely expand) that don't require the user to know about the details of the PyTorch dispatcher.

**Comment by ezyang** [comment]
We discussed this in person and resolved to use impl but remove mention of other dispatch keys as options.

@zou3519 do you want me to stamp this anyway and you get this in a follow up?

**Comment by zou3519** [comment]
> We discussed this in person and resolved to use impl but remove mention of other dispatch keys as options.
> 
> @zou3519 do you want me to stamp this anyway and you get this in a follow up?

No, let me open a new PR for this. We are also going to enable users passing in device names (e.g. "cpu", "cuda") via the impl APIs.

**Pull request closed without merging**